.maintain/gitlab/check_dependent_project.sh

```diff
@@ -0,0 +1,262 @@
+          fi
+        fi
+      ;;
+      *)
+        die "ERROR: Unknown state $next"
+      ;;
+    esac
+  done < <(cargo metadata --quiet --format-version=1 | jq -r '
+    . as $in |
+    paths(select(type=="string")) |
+    select(.[-1]=="source") as $source_path |
+    del($source_path[-1]) as $path |
+    [$in | getpath($path + ["name"]), getpath($path + ["source"])] |
+    .[]
+  ')
+  if [ -z "${last_line+_}" ]; then
+    die "No lines were read for cargo metadata of $PWD (some error probably occurred)"
+  fi
+
+  if [ "${crates_not_found[@]}" ]; then
+    echo -e "Errors during crate matching\n"
+    printf "Failed to detect our crate \"%s\" referenced in $target_name\n" "${crates_not_found[@]}"
+    echo -e "\nNote: this error generally happens if you have deleted or renamed a crate and did not update it in $target_name. Consider opening a companion pull request on $target_name and referencing it in this pull request's description like:\n$target_name companion: [your companion PR here]"
+    die "Check failed"
+  fi
+}
+
+patch_and_check_dependent() {
+  match_their_crates "$(basename "$PWD")"
+  diener patch --crates-to-patch "$this_repo_dir" "$this_repo_diener_arg" --path "Cargo.toml"
+  cargo check --all-targets --workspace
+}
+
+process_companion_pr() {
+  # e.g. https://github.com/paritytech/polkadot/pull/123
+  # or   polkadot#123
+  local companion_expr="$1"
+  if
+    [[ "$companion_expr" =~ ^https://github\.com/$org/([^/]+)/pull/([[:digit:]]+) ]] ||
+    [[ "$companion_expr" =~ ^$org/([^#]+)#([[:digit:]]+) ]] ||
+    [[ "$companion_expr" =~ ^([^#]+)#([[:digit:]]+) ]]; then
+    local companion_repo="${BASH_REMATCH[1]}"
+    local companion_pr_number="${BASH_REMATCH[2]}"
+    echo "Parsed companion_repo=$companion_repo and companion_pr_number=$companion_pr_number from $companion_expr (trying to match companion_repo=$dependent_repo)"
+  else
+    die "Companion PR description had invalid format or did not belong to organization $org: $companion_expr"
+  fi
+
+  if [ "$companion_repo" != "$dependent_repo" ]; then
+    return
+  fi
+
+  was_companion_found=true
+
+  read -d '\n' -r mergeable pr_head_ref pr_head_sha < <(curl \
+      -sSL \
+      -H "Authorization: token $GITHUB_TOKEN" \
+      "$gh_api/repos/$org/$companion_repo/pulls/$companion_pr_number" | \
+    jq -r "[
+      .mergeable // error(\"Companion $companion_expr is not mergeable\"),
+      .head.ref // error(\"Missing .head.ref from API data of $companion_expr\"),
+      .head.sha // error(\"Missing .head.sha from API data of $companion_expr\")
+    ] | .[]"
+  )
+
+  local expected_mergeable=true
+  if [ "$mergeable" != "$expected_mergeable" ]; then
+    die "Github API says $companion_expr is not mergeable (got $mergeable, expected $expected_mergeable)"
+  fi
+
+  git clone --depth 1 "https://github.com/$org/$companion_repo.git"
+  pushd "$companion_repo" >/dev/null
+  git fetch origin "pull/$companion_pr_number/head:$pr_head_ref"
+  git checkout "$pr_head_sha"
```

Comment:
this is building CI within CI, in my view it should be simplified.

Reply:
This is getting the companion's branch, which is not available here since we are in substrate and therefore we only have access to this repository. We need this step for patching the companion with `diener`.

It's what's being done in https://github.com/paritytech/substrate/blob/644c17eda766f392b787c8c0e01e25748d644b29/.maintain/gitlab/check_polkadot_companion_build.sh#L78 so it's at least not more complicated than the current approach.

I'm not noticing any extraneous steps there and it's as simple as I can make it right now.

Reply:
As you mentioned, it should be a part of a future overhaul.